lib/compile/validate/dataType.ts

```diff
@@ -92,28 +92,30 @@ function coerceData(it: SchemaObjCxt, types: JSONType[], coerceTo: JSONType[]):
         gen
           .elseIf(_`${dataType} == "number" || ${dataType} == "boolean"`)
           .assign(coerced, _`"" + ${data}`)
-          .elseIf(_`${data} === null`)
+          .elseIf(_`${data} === null && ${opts.coerceNull}`)
```

Comment:
a more efficient approach here and below would be not to check the option at the time of the validation, but to generate different code dependent on this option. So in this particular case, rather than changing the condition, the whole branch would simply be not included in the generated code.

Reply:
I could use some further clarification where to implement this. I haven't really had the time to go through the code in detail yet to fully understand where everything is done.

Reply:
that's exactly here, but rather than adding to the condition to the generated validation code, you could have this condition in the compilation code, e.g. in this case instead of:

``` javascript
gen
          .elseIf(_`${dataType} == "number" || ${dataType} == "boolean"`)
          .assign(coerced, _`"" + ${data}`)
          .elseIf(_`${data} === null`)
          .assign(coerced, _`""`)
```

you would have, for example:

``` javascript
gen
  .elseIf(_`${dataType} == "number" || ${dataType} == "boolean"`)
  .assign(coerced, _`"" + ${data}`)
if (opts.coerceNull) {
   gen
     .elseIf(_`${data} === null`)
     .assign(coerced, _`""`)
}
```

Just keep in mind so whatever code is inside ```_`..```` is executed during validation, you want as little as possible there, everything outside is the compilation code - given that options are static, no need to check them during validation - hope it makes sense.

It might be better to check the option first and if the data is null then skip adding all coercion checks to the validation code.

Reply:
I moved the `null` check before any `coerceTypes` usage so no validation is generated and the type error is reported. The added test succeeds but unsure if the implementation is to your liking.